File: src/Engine/EngineMath/BoundingSphere.cpp
#include "BoundingSphere.h"
#include "Matrix.h"
#include "EngineCommon.h"
#include "BoundingBox.h"
#include "BoundingFrustum.h"
#include "Plane.h"
#include "Ray.h"
#include <math.h>

BoundingSphere::BoundingSphere()
{
	Center = Vector3(0.0f,0.0f,0.0f);
	Radius = 0.0f;
};

BoundingSphere::BoundingSphere(const Vector3* center, FLOAT radius)
{
	Center = Vector3(center->X, center->Y, center->Z);
	Radius = radius;
};

BoundingSphere::~BoundingSphere()
{
};

BOOL BoundingSphere::Equals(const BoundingSphere* value1, const BoundingSphere* value2)
{
	return Vector3::Equals(&value1->Center, &value2->Center) && value1->Radius == value2->Radius;
};

void BoundingSphere::CreateMerged(const BoundingSphere *original, const BoundingSphere *additional, BoundingSphere *result)
{
	Vector3 Vector2;
    Vector3::Subtract(&additional->Center, &original->Center, &Vector2);
    FLOAT num = Vector3::Length(&Vector2);
    FLOAT radius = original->Radius;
    FLOAT num2 = additional->Radius;
    if ((radius + num2) >= num)
    {
        if ((radius - num2) >= num)
        {
			result = new BoundingSphere(&original->Center, original->Radius);
            return;
        }
        if ((num2 - radius) >= num)
        {
            result = new BoundingSphere(&additional->Center, additional->Radius);
            return;
        }
    }
    Vector3 vector;
	Vector3::Multiply(&Vector2, 1.0f / num, &vector);
    FLOAT num5 = GenericTools::Min(-radius, num - num2);;
	FLOAT num4 = (GenericTools::Max(radius, num + num2) - num5) * 0.5f;
	Vector3::Multiply(&vector, num4 + num5, &vector);
	Vector3::Add(&original->Center, &vector, &result->Center);
    result->Radius = num4;
};

void BoundingSphere::CreateFromBoundingBox(const BoundingBox *box, BoundingSphere *result)
{
	Vector3::Lerp(&box->Min, &box->Max, 0.5f, &result->Center);
    FLOAT num = Vector3::Distance(&box->Min, &box->Max);
    result->Radius = num * 0.5f;
};

void BoundingSphere::CreateFromPoints(INT arraySize, Vector3 points[], BoundingSphere *result)
{
    Vector3 Vector2;
    Vector3 vector5;
    Vector3 vector6;
    Vector3 vector7;
    Vector3 vector8;
    Vector3 vector9;
    
    Vector3 Vector4 = vector5 = vector6 = vector7 = vector8 = vector9 = points[0];
	for(int i = 1; i < arraySize; i++)
	{
		if (points[i].X < Vector4.X)
            Vector4 = points[i];
        if (points[i].X > vector5.X)
            vector5 = points[i];
        if (points[i].Y < vector6.Y)
            vector6 = points[i];
        if (points[i].Y > vector7.Y)
            vector7 = points[i];
        if (points[i].Z < vector8.Z)
            vector8 = points[i];
        if (points[i].Z > vector9.Z)
            vector9 = points[i];
	}
    
    FLOAT num5 = Vector3::Distance(&vector5, &Vector4);
    FLOAT num4 = Vector3::Distance(&vector7, &vector6);
	FLOAT num2 = Vector3::Distance(&vector9, &vector8);
	FLOAT num = 0;
    if (num5 > num4)
    {
        if (num5 > num2)
        {
            Vector3::Lerp(&vector5, &Vector4, 0.5f, &Vector2);
            num = num5 * 0.5f;
        }
        else
        {
            Vector3::Lerp(&vector9, &vector8, 0.5f, &Vector2);
            num = num2 * 0.5f;
        }
    }
    else if (num4 > num2)
    {
        Vector3::Lerp(&vector7, &vector6, 0.5f, &Vector2);
        num = num4 * 0.5f;
    }
    else
    {
        Vector3::Lerp(&vector9, &vector8, 0.5f, &Vector2);
        num = num2 * 0.5f;
    }
	for(int i = 0; i < arraySize; i++)
	{
		Vector3 Vector3;
        Vector3.X = points[i].X - Vector2.X;
        Vector3.Y = points[i].Y - Vector2.Y;
        Vector3.Z = points[i].Z - Vector2.Z;
        FLOAT num3 = Vector3::Length(&Vector3);
        if (num3 > num)
        {
            num = (num + num3) * 0.5f;
			Vector3::Multiply(&Vector3, 1.0f - (num / num3), &Vector3);
			Vector3::Add(&Vector2, &Vector3, &Vector2);
        }
	}
	result->Center.X = Vector2.X;
	result->Center.Y = Vector2.Y;
	result->Center.Z = Vector2.Z;
	result->Radius = num;
};

void BoundingSphere::CreateFromFrustum(const BoundingFrustum *frustum, BoundingSphere *result)
{
	BoundingSphere::CreateFromPoints(BoundingFrustum::CORNER_COUNT, frustum->pCornerArray, result);
};
	
BOOL BoundingSphere::Intersects(const BoundingSphere *sphere, const BoundingBox *box)
{
	FLOAT num = 0;
    Vector3 vector(0);
    Vector3::Clamp(&sphere->Center, &box->Min, &box->Max, &vector);
    num = Vector3::DistanceSquared(&sphere->Center, &vector);
    return (num <= (sphere->Radius * sphere->Radius));
};

BOOL BoundingSphere::Intersects(const BoundingSphere *sphere, const BoundingFrustum *frustum)
{
    return BoundingFrustum::Intersects(frustum, sphere);
};

PlaneIntersectionType BoundingSphere::Intersects(const BoundingSphere *sphere, const Plane *plane)
{
	return Plane::Intersects(plane, sphere);
};

BOOL BoundingSphere::Intersects(const BoundingSphere *sphere, const Ray *ray, FLOAT &result)
{
	return Ray::Intersects(ray, sphere, result);
};

BOOL BoundingSphere::Intersects(const BoundingSphere *sphere1, const BoundingSphere *sphere2)
{
	FLOAT num3;
	num3 = Vector3::DistanceSquared(&sphere1->Center, &sphere2->Center);
	FLOAT radius = sphere1->Radius;
	FLOAT num = sphere2->Radius;
	if ((((radius * radius) + ((2.0f * radius) * num)) + (num * num)) <= num3)
	{
		return FALSE;
	}
	return TRUE;
};

ContainmentType BoundingSphere::Contains(const BoundingSphere *sphere, const BoundingBox *box)
{
	Vector3 vector(0);
	if (!BoundingBox::Intersects(box, sphere))
	{
		return CT_Disjoint;
	}

	FLOAT num = sphere->Radius * sphere->Radius;
	vector.X = sphere->Center.X - box->Min.X;
	vector.Y = sphere->Center.Y - box->Max.Y;
	vector.Z = sphere->Center.Z - box->Max.Z;
	if (Vector3::LengthSquared(&vector) > num)
	{
		return CT_Intersects;
	}
	vector.X = sphere->Center.X - box->Max.X;
	vector.Y = sphere->Center.Y - box->Max.Y;
	vector.Z = sphere->Center.Z - box->Max.Z;
	if (Vector3::LengthSquared(&vector) > num)
	{
		return CT_Intersects;
	}
	vector.X = sphere->Center.X - box->Max.X;
	vector.Y = sphere->Center.Y - box->Min.Y;
	vector.Z = sphere->Center.Z - box->Max.Z;
	if (Vector3::LengthSquared(&vector) > num)
	{
		return CT_Intersects;
	}
	vector.X = sphere->Center.X - box->Min.X;
	vector.Y = sphere->Center.Y - box->Min.Y;
	vector.Z = sphere->Center.Z - box->Max.Z;
	if (Vector3::LengthSquared(&vector) > num)
	{
		return CT_Intersects;
	}
	vector.X = sphere->Center.X - box->Min.X;
	vector.Y = sphere->Center.Y - box->Max.Y;
	vector.Z = sphere->Center.Z - box->Min.Z;
	if (Vector3::LengthSquared(&vector) > num)
	{
		return CT_Intersects;
	}
	vector.X = sphere->Center.X - box->Max.X;
	vector.Y = sphere->Center.Y - box->Max.Y;
	vector.Z = sphere->Center.Z - box->Min.Z;
	if (Vector3::LengthSquared(&vector) > num)
	{
		return CT_Intersects;
	}
	vector.X = sphere->Center.X - box->Max.X;
	vector.Y = sphere->Center.Y - box->Min.Y;
	vector.Z = sphere->Center.Z - box->Min.Z;
	if (Vector3::LengthSquared(&vector) > num)
	{
		return CT_Intersects;
	}
	vector.X = sphere->Center.X - box->Min.X;
	vector.Y = sphere->Center.Y - box->Min.Y;
	vector.Z = sphere->Center.Z - box->Min.Z;
	if (Vector3::LengthSquared(&vector) > num)
	{
		return CT_Intersects;
	}

	return CT_Contains;
};

ContainmentType BoundingSphere::Contains(const BoundingSphere *sphere, const BoundingFrustum *frustum)
{
    if (!BoundingFrustum::Intersects(frustum, sphere))
    {
        return CT_Disjoint;
    }

    FLOAT num2 = sphere->Radius * sphere->Radius;
	for (INT i = 0; i < BoundingFrustum::CORNER_COUNT; i++)
	{
        Vector3 vector(0);
        vector.X = frustum->pCornerArray[i].X - sphere->Center.X;
        vector.Y = frustum->pCornerArray[i].Y - sphere->Center.Y;
        vector.Z = frustum->pCornerArray[i].Z - sphere->Center.Z;
        if (Vector3::LengthSquared(&vector) > num2)
        {
            return CT_Intersects;
        }
	}

    return CT_Contains;
};

ContainmentType BoundingSphere::Contains(const BoundingSphere *sphere, const Vector3 *point)
{
	if (Vector3::DistanceSquared(point, &sphere->Center) >= (sphere->Radius * sphere->Radius))
	{
		return CT_Disjoint;
	}
	return CT_Contains;
};

ContainmentType BoundingSphere::Contains(const BoundingSphere *sphere1, const BoundingSphere *sphere2)
{
	FLOAT num3;
    num3 = Vector3::Distance(&sphere1->Center, &sphere2->Center);
    FLOAT radius = sphere1->Radius;
    FLOAT num = sphere2->Radius;
    if ((radius + num) < num3)
    {
        return CT_Disjoint;
    }
    if ((radius - num) < num3)
    {
        return CT_Intersects;
    }
    return CT_Contains;

};

void BoundingSphere::SupportMapping(const BoundingSphere *sphere, const Vector3 *v, Vector3 *result)
{
	FLOAT num2 = Vector3::Length(v);
    FLOAT num = sphere->Radius / num2;
    result->X = 0;
	result->Y = 0;
	result->Z = 0;

    result->X = sphere->Center.X + (v->X * num);
    result->Y = sphere->Center.Y + (v->Y * num);
    result->Z = sphere->Center.Z + (v->Z * num);
};

void BoundingSphere::Transform(const BoundingSphere *sphere, const Matrix *matrix, BoundingSphere *result)
{
	Vector3::Transform(&sphere->Center, matrix, &result->Center);
    FLOAT num4 = ((matrix->M11 * matrix->M11) + (matrix->M12 * matrix->M12)) + (matrix->M13 * matrix->M13);
    FLOAT num3 = ((matrix->M21 * matrix->M21) + (matrix->M22 * matrix->M22)) + (matrix->M23 * matrix->M23);
    FLOAT num2 = ((matrix->M31 * matrix->M31) + (matrix->M32 * matrix->M32)) + (matrix->M33 * matrix->M33);
    FLOAT num = GenericTools::Max(num4, GenericTools::Max(num3, num2));
    result->Radius = sphere->Radius * (sqrt(num));
};

void BoundingSphere::Copy(const BoundingSphere *source, BoundingSphere *destination)
{
	Vector3::Copy(&source->Center, &destination->Center);
	destination->Radius = source->Radius;
};


BOOL BoundingSphere::operator ==(const BoundingSphere b) const
{
	return BoundingSphere::Equals(this, &b);
};

BOOL BoundingSphere::operator !=(const BoundingSphere b) const
{
	if (!(this->Center != b.Center))
    {
        return (this->Radius != b.Radius);
    }
    return true;
};

File: src/ActionEditor/ACFramework/Math/BoundingSphere.cs
using System;
using System.Net;
using System.Globalization;
using System.Collections.Generic;

namespace ACFramework
{
    public struct BoundingSphere : IEquatable<BoundingSphere>
    {
        public Vector3 Center;
        public float Radius;
        public BoundingSphere(Vector3 center, float radius)
        {
            if (radius < 0f)
            {
                throw new ArgumentException("Raio menor q zero");
            }
            this.Center = center;
            this.Radius = radius;
        }

        public bool Equals(BoundingSphere other)
        {
            return ((this.Center == other.Center) && (this.Radius == other.Radius));
        }

        public override bool Equals(object obj)
        {
            bool flag = false;
            if (obj is BoundingSphere)
            {
                flag = this.Equals((BoundingSphere)obj);
            }
            return flag;
        }

        public override int GetHashCode()
        {
            return (this.Center.GetHashCode() + this.Radius.GetHashCode());
        }

        public override string ToString()
        {
            CultureInfo currentCulture = CultureInfo.CurrentCulture;
            return string.Format(currentCulture, "{{Center:{0} Radius:{1}}}", new object[] { this.Center.ToString(), this.Radius.ToString(currentCulture) });
        }

        public static BoundingSphere CreateMerged(BoundingSphere original, BoundingSphere additional)
        {
            BoundingSphere sphere;
            Vector3 vector2;
            Vector3.Subtract(ref additional.Center, ref original.Center, out vector2);
            float num = vector2.Length();
            float radius = original.Radius;
            float num2 = additional.Radius;
            if ((radius + num2) >= num)
            {
                if ((radius - num2) >= num)
                {
                    return original;
                }
                if ((num2 - radius) >= num)
                {
                    return additional;
                }
            }
            Vector3 vector = (Vector3)(vector2 * (1f / num));
            float num5 = MathHelper.Min(-radius, num - num2);
            float num4 = (MathHelper.Max(radius, num + num2) - num5) * 0.5f;
            sphere.Center = original.Center + ((Vector3)(vector * (num4 + num5)));
            sphere.Radius = num4;
            return sphere;
        }

        public static void CreateMerged(ref BoundingSphere original, ref BoundingSphere additional, out BoundingSphere result)
        {
            Vector3 vector2;
            Vector3.Subtract(ref additional.Center, ref original.Center, out vector2);
            float num = vector2.Length();
            float radius = original.Radius;
            float num2 = additional.Radius;
            if ((radius + num2) >= num)
            {
                if ((radius - num2) >= num)
                {
                    result = original;
                    return;
                }
                if ((num2 - radius) >= num)
                {
                    result = additional;
                    return;
                }
            }
            Vector3 vector = (Vector3)(vector2 * (1f / num));
            float num5 = MathHelper.Min(-radius, num - num2);
            float num4 = (MathHelper.Max(radius, num + num2) - num5) * 0.5f;
            result.Center = original.Center + ((Vector3)(vector * (num4 + num5)));
            result.Radius = num4;
        }

        public static BoundingSphere CreateFromBoundingBox(BoundingBox box)
        {
            float num;
            BoundingSphere sphere;
            Vector3.Lerp(ref box.Min, ref box.Max, 0.5f, out sphere.Center);
            Vector3.Distance(ref box.Min, ref box.Max, out num);
            sphere.Radius = num * 0.5f;
            return sphere;
        }

        public static void CreateFromBoundingBox(ref BoundingBox box, out BoundingSphere result)
        {
            float num;
            Vector3.Lerp(ref box.Min, ref box.Max, 0.5f, out result.Center);
            Vector3.Distance(ref box.Min, ref box.Max, out num);
            result.Radius = num * 0.5f;
        }

        public static BoundingSphere CreateFromPoints(IEnumerable<Vector3> points)
        {
            float num;
            float num2;
            Vector3 vector2;
            float num4;
            float num5;
            BoundingSphere sphere;
            Vector3 vector5;
            Vector3 vector6;
            Vector3 vector7;
            Vector3 vector8;
            Vector3 vector9;
            if (points == null)
            {
                throw new ArgumentNullException("points");
            }
            IEnumerator<Vector3> enumerator = points.GetEnumerator();
            if (!enumerator.MoveNext())
            {
                throw new ArgumentException(FrameworkResources.BoundingSphereZeroPoints);
            }
            Vector3 vector4 = vector5 = vector6 = vector7 = vector8 = vector9 = enumerator.Current;
            foreach (Vector3 vector in points)
            {
                if (vector.X < vector4.X)
                {
                    vector4 = vector;
                }
                if (vector.X > vector5.X)
                {
                    vector5 = vector;
                }
                if (vector.Y < vector6.Y)
                {
                    vector6 = vector;
                }
                if (vector.Y > vector7.Y)
                {
                    vector7 = vector;
                }
                if (vector.Z < vector8.Z)
                {
                    vector8 = vector;
                }
                if (vector.Z > vector9.Z)
                {
                    vector9 = vector;
                }
            }
            Vector3.Distance(ref vector5, ref vector4, out num5);
            Vector3.Distance(ref vector7, ref vector6, out num4);
            Vector3.Distance(ref vector9, ref vector8, out num2);
            if (num5 > num4)
            {
                if (num5 > num2)
                {
                    Vector3.Lerp(ref vector5, ref vector4, 0.5f, out vector2);
                    num = num5 * 0.5f;
                }
                else
                {
                    Vector3.Lerp(ref vector9, ref vector8, 0.5f, out vector2);
                    num = num2 * 0.5f;
                }
            }
            else if (num4 > num2)
            {
                Vector3.Lerp(ref vector7, ref vector6, 0.5f, out vector2);
                num = num4 * 0.5f;
            }
            else
            {
                Vector3.Lerp(ref vector9, ref vector8, 0.5f, out vector2);
                num = num2 * 0.5f;
            }
            foreach (Vector3 vector10 in points)
            {
                Vector3 vector3 = Vector3.Zero;
                vector3.X = vector10.X - vector2.X;
                vector3.Y = vector10.Y - vector2.Y;
                vector3.Z = vector10.Z - vector2.Z;
                float num3 = vector3.Length();
                if (num3 > num)
                {
                    num = (num + num3) * 0.5f;
                    vector2 += (Vector3)((1f - (num / num3)) * vector3);
                }
            }
            sphere.Center = vector2;
            sphere.Radius = num;
            return sphere;
        }

        public bool Intersects(BoundingBox box)
        {
            float num;
            Vector3 vector;
            Vector3.Clamp(ref this.Center, ref box.Min, ref box.Max, out vector);
            Vector3.DistanceSquared(ref this.Center, ref vector, out num);
            return (num <= (this.Radius * this.Radius));
        }

        public void Intersects(ref BoundingBox box, out bool result)
        {
            float num;
            Vector3 vector;
            Vector3.Clamp(ref this.Center, ref box.Min, ref box.Max, out vector);
            Vector3.DistanceSquared(ref this.Center, ref vector, out num);
            result = num <= (this.Radius * this.Radius);
        }

        public float? Intersects(Ray ray)
        {
            return ray.Intersects(this);
        }

        public void Intersects(ref Ray ray, out float? result)
        {
            ray.Intersects(ref this, out result);
        }

        public bool Intersects(BoundingSphere sphere)
        {
            float num3;
            Vector3.DistanceSquared(ref this.Center, ref sphere.Center, out num3);
            float radius = this.Radius;
            float num = sphere.Radius;
            if ((((radius * radius) + ((2f * radius) * num)) + (num * num)) <= num3)
            {
                return false;
            }
            return true;
        }

        public void Intersects(ref BoundingSphere sphere, out bool result)
        {
            float num3;
            Vector3.DistanceSquared(ref this.Center, ref sphere.Center, out num3);
            float radius = this.Radius;
            float num = sphere.Radius;
            result = (((radius * radius) + ((2f * radius) * num)) + (num * num)) > num3;
        }

        internal void SupportMapping(ref Vector3 v, out Vector3 result)
        {
            float num2 = v.Length();
            float num = this.Radius / num2;
            result = Vector3.Zero;
            result.X = this.Center.X + (v.X * num);
            result.Y = this.Center.Y + (v.Y * num);
            result.Z = this.Center.Z + (v.Z * num);
        }

        public BoundingSphere Transform(Matrix matrix)
        {
            BoundingSphere sphere = new BoundingSphere
            {
                Center = Vector3.Transform(this.Center, matrix)
            };
            float num4 = ((matrix.M11 * matrix.M11) + (matrix.M12 * matrix.M12)) + (matrix.M13 * matrix.M13);
            float num3 = ((matrix.M21 * matrix.M21) + (matrix.M22 * matrix.M22)) + (matrix.M23 * matrix.M23);
            float num2 = ((matrix.M31 * matrix.M31) + (matrix.M32 * matrix.M32)) + (matrix.M33 * matrix.M33);
            float num = Math.Max(num4, Math.Max(num3, num2));
            sphere.Radius = this.Radius * ((float)Math.Sqrt((double)num));
            return sphere;
        }

        public void Transform(ref Matrix matrix, out BoundingSphere result)
        {
            result.Center = Vector3.Transform(this.Center, matrix);
            float num4 = ((matrix.M11 * matrix.M11) + (matrix.M12 * matrix.M12)) + (matrix.M13 * matrix.M13);
            float num3 = ((matrix.M21 * matrix.M21) + (matrix.M22 * matrix.M22)) + (matrix.M23 * matrix.M23);
            float num2 = ((matrix.M31 * matrix.M31) + (matrix.M32 * matrix.M32)) + (matrix.M33 * matrix.M33);
            float num = Math.Max(num4, Math.Max(num3, num2));
            result.Radius = this.Radius * ((float)Math.Sqrt((double)num));
        }

        public static bool operator ==(BoundingSphere a, BoundingSphere b)
        {
            return a.Equals(b);
        }

        public static bool operator !=(BoundingSphere a, BoundingSphere b)
        {
            if (!(a.Center != b.Center))
            {
                return (a.Radius != b.Radius);
            }
            return true;
        }
    }
}
